ext/CUDAExt.jl

```diff
@@ -0,0 +1,47 @@
+module CUDAExt
+
+using GCPDecompositions, CUDA
+
+GCPDecompositions.gcp(
+    X::CuArray,
+    r,
+    loss = LeastSquaresLoss();
+    constraints = (),
+    algorithm = GCPAlgorithms.ALS(),
+) = _gcp(X, r, loss, constraints, algorithm)
+function _gcp(
+    X::CuArray{TX,N},
+    r,
+    loss::LeastSquaresLoss,
+    constraints::Tuple{},
+    algorithm::GCPAlgorithms.ALS,
+) where {TX<:Real,N}
+    T = promote_type(TX, Float32)
+
+    # Random initialization
+    M0 = CPD(ones(T, r), rand.(T, size(X), r))
+    M0norm = sqrt(sum(abs2, M0[I] for I in CartesianIndices(size(M0))))
+    Xnorm = sqrt(mapreduce(x -> isnan(x) ? 0 : abs2(x), +, X, init=0f0))
+    for k in Base.OneTo(N)
+        M0.U[k] .*= (Xnorm / M0norm)^(1 / N)
+    end
+    λ, U = M0.λ, collect(M0.U)
+
+    # Move λ, U to gpu
+    λ = CuArray(λ)
+    U = [CuArray(U[i]) for i in 1:N]
+
+    # Inefficient but simple implementation
+    for _ in 1:algorithm.maxiters
+        for n in 1:N
+            V = reduce(.*, U[i]'U[i] for i in setdiff(1:N, n))
+            U[n] = GCPDecompositions.mttkrp(X, U, n) / V
+            λ = CuArray(CUDA.norm.(eachcol(U[n])))
```

Comment:
Looks like this step is much slower when having to move the data around devices:
```
julia> @btime norm.(eachcol(U_cpu[n]))
  1.408 μs (4 allocations: 192 bytes)
10-element Vector{Float32}:
 1.941741
 2.0177264
 1.904886
 1.636472
 1.4084961
 0.9433015
 1.0931745
 1.6607362
 1.273025
 1.8184978

julia> @btime CUDA.@sync CuArray(norm.(eachcol(U_gpu[n])))
  440.045 μs (173 allocations: 7.38 KiB)
10-element CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}:
 1.9417399
 2.0177276
 1.9048846
 1.6364713
 1.408497
 0.94329995
 1.0931759
 1.6607368
 1.2730248
 1.8184997
```

Reply:
Rewriting norm.(eachcol(U[n])) as vec(sqrt.(sum(abs2, U_gpu[n]; dims=1))) prevents data from being transferred back to CPU and gets a 5x speedup for GPU version, similar time for CPU:
```
julia> @btime norm.(eachcol(U_cpu[n]))
  1.364 μs (4 allocations: 192 bytes)
10-element Vector{Float32}:
 1.941741
 2.0177264
 1.904886
 1.636472
 1.4084961
 0.9433015
 1.0931745
 1.6607362
 1.273025
 1.8184978

julia> @btime vec(sqrt.(sum(abs2, U_cpu[n]; dims=1)))
  1.212 μs (6 allocations: 304 bytes)
10-element Vector{Float32}:
 1.9417411
 2.0177264
 1.904886
 1.6364719
 1.408496
 0.9433015
 1.0931743
 1.6607362
 1.273025
 1.8184978

julia> @btime CUDA.@sync CuArray(norm.(eachcol(U_gpu[n])))
  368.776 μs (173 allocations: 7.38 KiB)
10-element CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}:
 1.9417399
 2.0177276
 1.9048846
 1.6364713
 1.408497
 0.94329995
 1.0931759
 1.6607368
 1.2730248
 1.8184997

julia> @btime CUDA.@sync vec(sqrt.(sum(abs2, U_gpu[n]; dims=1)))
  67.685 μs (112 allocations: 5.52 KiB)
10-element CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}:
 1.9417399
 2.0177276
 1.9048846
 1.6364713
 1.408497
 0.9432999
 1.093176
 1.6607368
 1.2730248
 1.8184998

```